### PR TITLE
First-order meta-gradient every step on the MLP matrices: 3.295 val loss in 14.2m

### PR DESCRIPTION
**Pull request by dangxingyu**
Each optimizer step splits its 32-sequence device batch in two halves of 16. The gradient of the first
half is taken at the current weights, a temporary step of fixed L2 norm is applied along it on the MLP
matrices, the gradient of the second half is taken at that adapted point, the temporary step is undone,
and the optimizer receives the mean of the two gradients: the ordinary full-batch mean. The second
half's gradient at the adapted point is a first-order meta-gradient (first half = adaptation batch,
second half = evaluation batch). Everything is rank-local; the two halves cost about what the single
32-sequence forward/backward cost. The recipe's dropout, weight decay, and schedules are unchanged
from the baseline (the current record). The same episode, on all Muon matrices, is submitted separately for the 1-hour track (https://github.com/qlabs-eng/slowrun/pull/98).

Per step (the script's defaults):

```
g   = 1/2 * mean-grad(theta; sequences 1-16)        # first half of the device batch, on this rank
d   = g|P / ||g|P||                                  # P = the 48 MLP matrices (138M params)
theta_P -= eta * d,  eta = 0.5                       # constant step norm
g  += 1/2 * mean-grad(theta; sequences 17-32)       # second half at the adapted point
theta_P += eta * d                                   # restore: only the matrices this rank owns in the Muon step (the all-gather overwrites the rest)
optimizer.step(g)
```

Results on one 8xH100 80GB HBM3 node, val loss (training time):

| seed | baseline | this script |
| --- | --- | --- |
| 42 | 3.3095 (13.8 min) | 3.2954 (14.2 min) * |
| 43 | 3.3046 (13.9 min) | 3.2958 (14.6 min) |
| 44 | 3.3070 (14.0 min) | 3.2939 (14.3 min) |
| 45 | 3.3101 (13.8 min) | 3.2984 (14.3 min) |

\* the submitted run. Gain on the same seed: 0.014 / 0.009 / 0.013 / 0.012; fixed-seed run-to-run
noise is about 0.001-0.002. Published record: 3.307 at 14.1 min.

Timing: the episode adds 8-14 ms to a 270 ms step. The tiny timer also counts recompiles, and the
recipe's 1x -> 2x recurrence transition (step 1996) recompiled the episode's half-batch shape in the
first version of this script, which cost 17-150 s. The submitted script precompiles the half-batch
shape for every recurrence stage before the timer starts; in every listed run step 1996 took the normal
2x-stage step time. Seed 43's 14.6 min is a node that is 3 percent slower per step.

Script: `tiny/train.py` (the baseline script) with the episode added (+72 / -10 lines). Defaults reproduce
this entry; `--ttt-every 0` gives back the baseline loop. Reproduce:

```bash
torchrun --standalone --nproc_per_node=8 tiny/train.py --seed 42
# equivalent explicit flags: --ttt-every 1 --ttt-step-norm 0.5 --ttt-step-schedule const --ttt-plastic mlp_all
```

Training logs of every run in the table: https://gist.github.com/dangxingyu/d45291f9ee1e3953879d5f65dbcd57a0 (one `Result saved to <path>` line per log redacted).

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -78,6 +78,7 @@ The tiny track caps runs at a single 8xH100 node for at most 15 minutes.
 12 | 3.324 | Add 2x recurrence | 06/07/26 | 14.9 mins | [Script](https://github.com/qlabs-eng/slowrun/blob/a2c443e18f99d0f6e35c6013c752e636fe402213/tiny/train.py) | [@neel04](https://x.com/awesome_ruler_)
 13 | 3.315 | Add fp8 MTP + full-layer XSA | 06/08/26 | 13.9 mins | [Script](https://github.com/qlabs-eng/slowrun/blob/9732ac294aba594a1fbb3aff1bb93cfdf2ce6788/tiny/train.py) | [@Mister-dev-oss](https://github.com/Mister-dev-oss)
 14 | 3.307 | Merge 2x recurrence with latest changes | 06/14/26 | 14.1 mins | [Script](https://github.com/qlabs-eng/slowrun/blob/d1fbdb76386347f1097ed051f4bebaef2c17758d/tiny/train.py) | [@neel04](https://x.com/awesome_ruler_)
+15 | 3.295 | First-order meta-gradient every step on the MLP matrices (split half-batches, temporary step of norm 0.5); precompile the half-batch shape | 09/05/26 | 14.2 mins | [Script](https://github.com/dangxingyu/slowrun/blob/ttt-every-step-tiny/tiny/train.py) | [@dangxingyu](https://github.com/dangxingyu)
```
Comment (akshayvegesna) [inline README.md]: Remove, I'll add this after we merge

File: tiny/train.py [hunk 2]
```diff
@@ -126,6 +126,12 @@
                     help="MTP: number of future tokens predicted from each position (1 = plain next-token).")
 parser.add_argument("--mtp-anneal-frac", type=float, default=0.66,
                     help="Fraction of training over which the extra MTP heads decay to zero weight.")
+# --- every-step training-time-testing episode (default for this entry; --ttt-every 0 = entry 14) ---
+parser.add_argument("--seed", type=int, default=42)
+parser.add_argument("--ttt-every", type=int, default=1, help="episode every N steps (1 = every step, the record; 0 = off = entry 14)")
```
Comment (akshayvegesna) [inline tiny/train.py]: I think when people see TTT they will think test time training, perhaps we abbreviate as MG for metagradient

**Review by akshayvegesna — COMMENTED** [review]
This is great! Left minor comments, let's merge after those are updated.

A couple questions as well
- What have you tuned so far? Do you think it can be extended? 
- Why do you choose to do normalized gradient descent instead of SGD?

**Comment by dangxingyu** [comment]
Thanks for the review. Both branches are updated: the README row is removed and the `ttt` flags and identifiers are renamed to `mg` (`--mg-every`, `--mg-step-norm`, `--mg-step-schedule`, `--mg-plastic`).

**What we tuned** (one-hour track, seed 42, same-seed baseline = the record script at 3.2005; all numbers are the leaderboard quantity after the checkpoint average; run-to-run noise at a fixed seed is about 0.002)

Starting configuration: at every optimizer step, split the batch in two halves, take a temporary normalized step of norm 0.3 on the Muon matrices along the first half's gradient, take the second half's gradient at the adapted weights, restore, and hand the mean of the two gradients to Muon. That gives 3.1895 (-0.011).

| change | result |
|---|---|
| constant step norm 0.1 / 0.3 / 1.0 | 3.1931 / 3.1915 / 3.2088 |
| step norm 0.5·m(t), following the warmdown LR multiplier | 3.1857 (-0.004 on top of 0.3) |
| matrices that take the temporary step: all Muon matrices / MLP matrices only | 3.1915 / 3.1912 |
| adapt on the current batch and test on the next batch, instead of splitting the batch | 3.1948 at every step, at 2x the step time; every arm of a 17-arm sweep of that pairing (step geometry, norm 0.03 to 1, matrix subset, schedule, weight decay) lands within -0.001 to +0.009 of the baseline |
| how often: every 16 / 8 / 4 / 2 / 1 steps | gains 0.002 to 0.003 for 16, 8, 4; 0.005 for 2; 0.009 for every step |
| direction at the submitted setting: descent / random of the same norm / ascent (SAM-like) | 3.1811 / 3.1998 / 3.2035 |
| geometry of the temporary step: raw gradient, Muon-shaped (orthogonalized, 0.05x to 2x the live LR), sign, momentum-smoothed | all worse than normalized; Muon-shaped costs +0.001 to +0.008 plus 8 minutes of orthogonalization |
| normalize each rank's own gradient shard vs the all-reduced direction | 3.1895 vs 3.1915 / 3.1935 (two implementations), and 30 ms per step cheaper |
| recipe dropout 0.1 -> 0.05 / 0, with the temporary step in place | 3.1854 / 3.1843 (0 to 0.05 flat, 0.1 worse) |
| dropout 0.05 + step norm 0.5·m(t) | 3.1832 |
| + stochastic depth 0.05 -> 0 | 3.1811, the submitted configuration (3.1807 / 3.1814 for two restore implementations) |
| use only the adapted-point gradient, no averaging with the original | +0.021 |
| reuse the adapted-point gradient as the next step's gradient, so the extra work is free | +0.012 |
| unequal weighting of the two half gradients (6/2, 2/6, 1/7) | within 0.002 of the equal split |
| temporary steps only in the first half of training / only in the averaging window | keeps 0.007 / 0.003 of the 0.011 |

Neutral or harmful on top of that: peak LR, warmdown length, MTP weight, two temporary steps per optimizer step, a four-checkpoint averaging window. On the tiny track the same design uses the MLP matrices and norm 0.5; there, removing dropout costs 0.007 and halving it is neutral, so the tiny script keeps it.

On four seeds per track: 3.1814 to 3.1834 vs 3.2005 to 3.2034 for the record script on the same seeds (one hour), and 3.2939 to 3.2984 vs 3.3046 to 3.3101 (tiny).

**Why a normalized step rather than plain SGD**

Raw-gradient temporary steps were no better than the baseline at the step sizes we tried. The reason is scale: the half-batch gradient over these matrices is small (norm of order 1e-3) and its norm changes several-fold over training, so a raw step either moves the weights by a negligible distance or needs a step size that tracks the gradient norm. Normalizing makes the temporary step a fixed Euclidean distance eta, which is the quantity that matters for the lookahead (the second gradient is taken a fixed distance downhill from the current weights). That is why one value of eta is flat between 0.1 and 0.5 and the same setting works on both tracks. Preconditioned steps (Muon-shaped, sign) recover part of the gain, but only in a narrow step-size window.

**Comment by akshayvegesna** [comment]
Thanks!
